admin-dev/themes/new-theme/js/pages/tax-rules-group/TaxRulesGroupForm.js

```diff
@@ -0,0 +1,346 @@
+    $.each(response, function (index, value) {
+      $stateFormSelect.append($('<option></option>').attr('value', value).text(index));
+    });
+
+    if ($stateFormRow.hasClass('hidden-element')) {
+      $stateFormRow.removeClass('hidden-element');
+    }
+
+    $stateFormSelect.removeAttr('disabled');
+    $stateFormRow.fadeIn(fadeSpeed);
+  }
+
+  /**
+   * Removes form errors if new form data is loaded
+   *
+   * @private
+   */
+  _removeTaxRuleFormErrors() {
+    if ($(taxRulesGroupFormMap.taxRuleForm).find(taxRulesGroupFormMap.taxRuleFormErrorAlert).length) {
+      $(taxRulesGroupFormMap.taxRuleForm).find(taxRulesGroupFormMap.taxRuleFormErrorAlert).remove();
+    }
+
+    $(taxRulesGroupFormMap.taxRuleInvalidFeedbackContainer).remove();
+    $(taxRulesGroupFormMap.taxRuleErrorPopoverContent).remove();
+    $.each(taxRulesGroupFormMap.taxRuleFormPopoverErrorContainers, function (index, value) {
+      $(value).remove();
+    });
+  }
+
+  /**
+   * Fills tax rule form fields with response data
+   *
+   * @param data
+   *
+   * @private
+   */
+  _setTaxRuleInformation(data) {
+    const $countrySelect = $(taxRulesGroupFormMap.countrySelect);
+
+    let taxRule = null;
+    let states = [];
+
+    let countryId = 0;
+    let stateId = 0;
+    let zipCode = 0;
+    let behaviorId = 0;
+    let taxId = 0;
+    let description = '';
+
+    this._removeTaxRuleFormErrors();
+
+    if (data !== null) {
+      taxRule = data[0];
+      states = data[1];
+
+      behaviorId = taxRule.behavior_id.value;
+      countryId = taxRule.country_id.value;
+      zipCode = taxRule.zip_code;
+
+      stateId = taxRule.state_id === null ? 0 : taxRule.state_id.value;
+      taxId = taxRule.tax_id === null ? 0 : taxRule.tax_id.value;
+      description = taxRule.description === null ? 0 : taxRule.description;
+    }
+
+    $countrySelect.val(countryId);
+
+    this._resolveStateSelectVisibility(states, 0);
+
+    $(taxRulesGroupFormMap.stateFormSelect).val(stateId);
+    $(taxRulesGroupFormMap.zipCodeInput).val(zipCode);
+    $(taxRulesGroupFormMap.behaviorSelect).val(behaviorId);
+    $(taxRulesGroupFormMap.taxSelect).val(taxId);
+    $(taxRulesGroupFormMap.descriptionInput).val(description);
+  }
+
+  /**
+   * Enables loading spinner for tax rule form
+   *
+   * @private
+   */
+  _enableSpinner() {
+    if (!$('#spinner-block').length) {
+      const $taxRuleForm = $(taxRulesGroupFormMap.taxRuleForm);
+
+      $(taxRulesGroupFormMap.taxRulesHiddenContent).hide();
+      $taxRuleForm.find('.card').append(this._getSpinnerBlock());
+    }
+  }
+
+  /**
+   * Disables loading spinner for tax rule form
+   *
+   * @private
+   */
+  _disableSpinner() {
+    if ($('#spinner-block').length) {
+      $('#spinner-block').remove();
+      $(taxRulesGroupFormMap.taxRulesHiddenContent).fadeIn();
+    }
+  }
+
+  /**
+   * Return html loading spinner element
+   *
+   * @returns {string}
+   *
+   * @private
+   */
+  _getSpinnerBlock() {
```

Comment:
@PierreRambaud I'm bad at frontend so I seek your help. How could we avoid this html-in-js ? Put the spinner item hidden in the template ?

Reply:
Build it with JS. :trollface:

Reply:
Build it with JS. :trollface: